erasure-coding/src/lib.rs

```diff
@@ -0,0 +1,409 @@
+	NonUniformChunks,
+	/// An uneven byte-length of a shard is not valid for GF(2^16) encoding.
+	UnevenLength,
+	/// Chunk index out of bounds.
+	ChunkIndexOutOfBounds(usize, usize),
+	/// Bad payload in reconstructed bytes.
+	BadPayload,
+	/// Invalid branch proof.
+	InvalidBranchProof,
+	/// Branch out of bounds.
+	BranchOutOfBounds,
+}
+
+struct CodeParams {
+	data_shards: usize,
+	parity_shards: usize,
+}
+
+impl CodeParams {
+	// the shard length needed for a payload with initial size `base_len`.
+	fn shard_len(&self, base_len: usize) -> usize {
+		(base_len / self.data_shards) + (base_len % self.data_shards)
+	}
+
+	fn make_shards_for(&self, payload: &[u8]) -> Vec<WrappedShard> {
+		let shard_len = self.shard_len(payload.len());
+		let mut shards = vec![
+			WrappedShard::new(vec![0; shard_len + 4]);
+			self.data_shards + self.parity_shards
+		];
+
+		for (data_chunk, blank_shard) in payload.chunks(shard_len).zip(&mut shards) {
+			let blank_shard: &mut [u8] = blank_shard.as_mut();
+			let (len_slice, blank_shard) = blank_shard.split_at_mut(4);
+			let len = ::std::cmp::min(data_chunk.len(), blank_shard.len());
+
+			// prepend the length to each data shard. this will tell us how much
+			// we need to read.
+			//
+			// this is necessary because we are doing RS encoding with 16-bit words,
+			// but the payload is a byte-slice. We need to know how much data
+			// to read from each shard when reconstructing.
+			//
+			// TODO: could be done more efficiently by pushing extra bytes onto the
+			// end. https://github.com/paritytech/polkadot/issues/88
+			(len as u32).using_encoded(|s| {
+				len_slice.copy_from_slice(s)
+			});
+
+			// fill the empty shards with the corresponding piece of the payload,
+			// zero-padded to fit in the shards.
+			blank_shard[..len].copy_from_slice(&data_chunk[..len]);
+		}
+
+		shards
+	}
+
+	// make a reed-solomon instance.
+	fn make_encoder(&self) -> ReedSolomon {
+		ReedSolomon::new(self.data_shards, self.parity_shards)
+			.expect("this struct is not created with invalid shard number; qed")
+	}
+}
+
+fn code_params(n_validators: usize) -> Result<CodeParams, Error> {
+	if n_validators > MAX_VALIDATORS { return Err(Error::TooManyValidators) }
+	if n_validators == 0 { return Err(Error::EmptyValidators) }
+
+	let n_faulty = n_validators.saturating_sub(1) / 3;
+	let n_good = n_validators - n_faulty;
+
+	Ok(CodeParams {
+		data_shards: n_faulty + 1,
+		parity_shards: n_good - 1,
+	})
+}
+
+/// Obtain erasure-coded chunks, one for each validator.
+///
+/// Works only up to 256 validators, and `n_validators` must be non-zero.
+pub fn obtain_chunks(n_validators: usize, block_data: &BlockData, extrinsic: &Extrinsic)
+	-> Result<Vec<Vec<u8>>, Error>
+{
+	let params  = code_params(n_validators)?;
+	let encoded = (block_data, extrinsic).encode();
+
+	if encoded.is_empty() {
+		return Err(Error::BadPayload);
+	}
+
+	let mut shards = params.make_shards_for(&encoded[..]);
```

Comment:
Ahh okay no reason in `Cow`s here, even if the ones outside my house are cute.  ;)